repo-agent/repowatch/controllers/repowatch_controller.go

```diff
@@ -54,6 +56,55 @@ var seededRand = rand.New(
 
 type githubClientFactory func(ctx context.Context, k8sClient client.Client, repoWatch *reviewv1alpha1.RepoWatch) (*github.Client, map[string]string, error)
 
+// PersistingTokenSource wraps an oauth2.TokenSource and persists the token to a Kubernetes secret when it changes.
+type PersistingTokenSource struct {
+	Source     oauth2.TokenSource
+	K8sClient  client.Client
+	SecretName string
+	Namespace  string
+}
+
+func (s *PersistingTokenSource) Token() (*oauth2.Token, error) {
+	t, err := s.Source.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	// Persist the token if it has changed
+	go func(token *oauth2.Token) {
+		ctx := context.Background()
+		secret := &corev1.Secret{}
+		if err := s.K8sClient.Get(ctx, types.NamespacedName{Name: s.SecretName, Namespace: s.Namespace}, secret); err != nil {
+			log.Log.Error(err, "failed to get secret for token update", "secret", s.SecretName)
+			return
+		}
+
+		if secret.Data == nil {
+			secret.Data = make(map[string][]byte)
+		}
+
+		// Check if token changed
+		currentPAT := string(secret.Data["pat"])
+		if currentPAT == token.AccessToken {
+			return
+		}
+
+		secret.Data["pat"] = []byte(token.AccessToken)
+		if token.RefreshToken != "" {
+			secret.Data["refresh_token"] = []byte(token.RefreshToken)
+		}
+		if !token.Expiry.IsZero() {
+			secret.Data["expiry"] = []byte(token.Expiry.Format(time.RFC3339))
+		}
+
+		if err := s.K8sClient.Update(ctx, secret); err != nil {
+			log.Log.Error(err, "failed to update secret with new token", "secret", s.SecretName)
```

Comment:
The error message "waiting for user login to populate github token" is helpful, but the reconciler will likely retry immediately, leading to a tight loop of log spam and wasted CPU cycles. It would be better to return a `ctrl.Result` with a `RequeueAfter` duration (e.g., 1 minute) to gracefully wait for the user to log in without overwhelming the controller.